src/CropManager.sol

```diff
@@ -0,0 +1,161 @@
+    event Deny(address indexed);
+    event SetImplementation(address indexed);
+
+    constructor() public {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    function rely(address usr) public auth { wards[usr] = 1; emit Rely(msg.sender); }
+    function deny(address usr) public auth { wards[usr] = 0; emit Deny(msg.sender); }
+    modifier auth { require(wards[msg.sender] == 1, "CropManager/non-authed"); _; }
+
+    function setImplementation(address implementation_) external auth {
+        implementation = implementation_;
+        emit SetImplementation(implementation_);
+    }
+
+    fallback() external {
+        address _impl = implementation;
+        require(_impl != address(0));
+
+        assembly {
+            let ptr := mload(0x40)
+            calldatacopy(ptr, 0, calldatasize())
+            let result := delegatecall(gas(), _impl, ptr, calldatasize(), 0, 0)
+            let size := returndatasize()
+            returndatacopy(ptr, 0, size)
+
+            switch result
+            case 0 { revert(ptr, size) }
+            default { return(ptr, size) }
+        }
+    }
+}
+
+contract CropManagerImp {
+    bytes32 slot0;
+    mapping (address => address) proxy;  // UrnProxy per user
+
+    address public immutable vat;
+    constructor(address vat_) public {
+        vat = vat_;
+    }
+
+    function getOrCreateProxy(address usr) public returns (address urp) {
+        urp = proxy[usr];
+        if (urp == address(0)) {
+            urp = proxy[usr] = address(new UrnProxy(address(vat), usr));
+        }
+    }
+
+    function join(address crop, address usr, uint256 val) external {
+        TokenLike(CropLike(crop).gem()).transferFrom(msg.sender, address(this), val);
+        TokenLike(CropLike(crop).gem()).approve(crop, val);
+        CropLike(crop).join(getOrCreateProxy(usr), usr, val);
+    }
+
+    function exit(address crop, address usr, uint256 val) external {
+        address urp = proxy[msg.sender];
+        require(urp != address(0), "CropManager/non-existing-urp");
+        CropLike(crop).exit(urp, usr, val);
+    }
+
+    function flee(address crop) external {
+        address urp = proxy[msg.sender];
+        require(urp != address(0), "CropManager/non-existing-urp");
+        CropLike(crop).flee(urp, msg.sender);
+    }
+
+    function frob(address crop, address u, address v, address w, int256 dink, int256 dart) external {
+        require(u == msg.sender && v == msg.sender && w == msg.sender, "CropManager/not-allowed");
+
+        // Note: This simplification only works because of the u == v == msg.sender restriction above
+        address urp = getOrCreateProxy(u);
+        VatLike(vat).frob(CropLike(crop).ilk(), urp, urp, w, dink, dart);
+    }
+
+    function flux(address crop, address src, address dst, uint256 wad) external {
+        require(src == msg.sender, "CropManager/not-allowed");
+
+        address surp = getOrCreateProxy(src);
+        address durp = getOrCreateProxy(dst);
```

Comment:
I'm a bit worried here someone can pass the proxy as the `dst` instead of the real user address and the funds get lost.

Reply:
Starting to think that maybe the proxy should be created first and not do it via `getOrCreateProxy`.

Reply:
IMO it doesn't make sense to create it for `surp`--that proxy should already exist. I don't feel strongly about `durp`--it doesn't seem a very likely mistake to send to the proxy instead of the user. Also it might actually be convenient to have the proxy created automatically for the recipient.

Reply:
If you guys don't see a risk there, then it's fine as it is. Not very strong feeling about it tbh.

Reply:
I mean I definitely think it's nonsensical to create a proxy for `src`--has similar issues to what I outlined with creating a proxy in `exit` or `flee`.

Reply:
Yes I was meaning just for the `dst` case, but let's leave it as it is.